### PR TITLE
Only emit circular dependency warning for owned thread shields

### PR DESCRIPTION
**Pull request by casperisfine**
[Bug #19415]

If multiple threads attemps to load the same file concurrently it's not a circular dependency issue.

So we check that the existing ThreadShield is owner by the current fiber before warning about circular dependencies.

FYI: @fxn 

NB: I'm not happy about having to rely on `sleep` for the test. I'm still looking for a better solution.

### CODE REVIEW

**Comment by casperisfine** [comment]
> NB: I'm not happy about having to rely on `sleep` for the test. I'm still looking for a better solution.

Of course the second I submitted, a solution hit me... Should be more acceptable now.

**Comment by fxn** [comment]
Can't be any faster :).

**Review by jeremyevans — APPROVED** [review]
I'm not familiar with the thread shield code, but this looks OK.

File: spec/ruby/fixtures/code/concurrent_require_fixture.rb
Comment (eregon) [inline spec/ruby/fixtures/code/concurrent_require_fixture.rb:2]: Why the `0.1` and not just `.join`?

**Comment by eregon** [comment]
The semantic change sounds good to me, in TruffleRuby we only warn if the per-file lock is `lock.isHeldByCurrentThread()` which is the same as "owned by the same Fiber".

File: spec/ruby/fixtures/code/concurrent_require_fixture.rb
Comment (eregon) [inline spec/ruby/fixtures/code/concurrent_require_fixture.rb:2]:
```suggestion
Thread.new { ScratchPad.recorded.require(__FILE__) }.join
```

Comment (casperisfine) [reply spec/ruby/fixtures/code/concurrent_require_fixture.rb:2]: Because that would cause a deadlock. The `ThreadShield` is help by the main thread, so that subthread will lock on the `require` call until the mainthread is done requiring.

Comment (eregon) [reply spec/ruby/fixtures/code/concurrent_require_fixture.rb:1]: Ah yes, right.
We need to fix that spec then, because it will leak a Thread. We should `join` in the caller to make sure that Thread doesn't live longer than the spec example.
You can use `CHECK_LEAKS=true` to detect such resources leaking between spec examples.

Also this will then wait 100ms just to ensure the thread starts, that's very slow for a ruby/spec spec, it's a lot more efficient to rely on `Thread#stop?` (or status) instead as other specs do.

Comment (byroot) [reply spec/ruby/fixtures/code/concurrent_require_fixture.rb:1]: Hum, good point, I'll improve that tomorrow (sorry for not waiting on your second review)

**Pull request merged**
